keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,415 @@
+backed by PMEM and provided by
+[PMEM-CSI](https://github.com/intel/pmem-csi). But PMEM is a resource
+that is local to a node and thus the scheduler has to be aware whether
+enough of it is available on a node before assigning a pod to it.
+
+
+### Caching remaining capacity via the API server
+
+CSI defines the GetCapacity RPC call for the controller service to
+report information about the current capacity but it's not accessible
+in the Kubernetes scheduler. A new API type which represents capacity
+is introduced to solve this (details below).
+
+The CSI driver deployment is responsible for creating and updating
+objects of that type. The external-provisioner gets extended to handle
+this, so CSI drivers only need to implement the GetCapacity call. Each
+capacity value has a lifetime chosen by the creator. Removal of
+expired values is left to the Kubernetes scheduler. This avoids the
+issue that two different processes both try the same operation, which
+would be redundant work. Making the Kubernetes scheduler responsible
+for this has the advantage that it also works when the driver crashes
+or gets de-installed.
+
+The Kubernetes scheduler monitors those objects and uses that
+information when it comes to making scheduling decisions. If the
+scheduler has no current capacity information, it proceeds without it,
+so nothing changes for CSI driver deployments that do not support
+capacity tracking.
+
+This is a problem for deployments that do support it, but haven’t been
+able to provide the necessary information yet. TODO: extend
+CSIDriverInfo with a field that tells the scheduler to retry later
+when information is missing? Avoid the situation by (Somehow? How?!)
+always providing the information before registering the driver -
+probably doesn’t work for PVCs with late binding because the pod gets
+scheduled even if the driver isn’t installed (which also is a problem
+for the CSIDriverInfo approach…).
+
+### API
+
+All Kubernetes API changes are under the `CSIStorageCapacity` feature
+gate.
+
+#### Storage capacity
+
+The prefix “CSI” is used to avoid confusion with other types that have
+a similar name and because some aspects are specific to CSI.
+
+```
+// CSIStorageInfo represents information about storage provided by a certain CSI driver,
+// like for example current capacity. As in CSIDriverInfo, the name of the CSIStorageInfo
+// objects is the same as the driver name. Unlike CSIDriverInfo, these objects are
+// created dynamically and get updated regularly when there are changes.
+type CSIStorageInfo struct {
+    metav1.TypeMeta
+    metav1.ObjectMeta
+
+   // The actual information may depend on the storage class and therefore
+   // is provided separately for each storage class that uses the driver.
+   // The empty class name is valid and represents information relevant for
+   // ephemeral inline volumes as those don't use storage classes.
+   Info: map[string]CSIStorageClassInfo
```

Comment:
https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#lists-of-named-subobjects-preferred-over-maps

-> will replace with a list.

Reply:
I'll update the KEP next week. In the meantime let me share two alternatives:

Stand-alone object:
```
// CSIStorage represents information about storage provided by a certain CSI driver,
// like for example current capacity. As in CSIDriver, the name of the CSIStorage
// objects is the same as the driver name. Unlike CSIDriver, these objects are
// created dynamically and get updated regularly when there are changes.
type CSIStorage struct {
	metav1.TypeMeta
	metav1.ObjectMeta

	// The actual information may depend on the storage class and therefore
	// is provided separately for each storage class that uses the driver.
	Info []CSIStorageByClass
}

// CSIStorageByClass contains information for one particular storage class
// of a CSI driver.
type CSIStorageByClass struct {
	// Storage class name, <all>, <ephemeral>.
	StorageClassName string

	// A CSI driver may allocate storage from one or more pools with different
	// attributes. Entries must have unique names inside this list.
	Pools []CSIStoragePool
}

// CSIStoragePoolInfo identifies one particular storage pool and
// stores the corresponding attributes.
type CSIStoragePool struct {
	// The name is some user-friendly identifier for this entry.
	Name string

	// NodeTopology can be used to describe a storage pool that is available
	// only for certain nodes in the cluster. If not set, the pool is consider
	// to be available from all nodes.
	NodeTopology *api.NodeSelector

	// Capacity is the size of the largest volume that currently can
	// be created. This is a best-effort guess and even volumes
	// of that size might not get created successfully.
	Capacity *resource.Quantity

	// ExpiryTime is the absolute time at which this entry becomes obsolete.
	// When not set, the entry is valid forever.
	ExpiryTime *metav1.Time
}
```

I'm not entirely happy with the names in that version (`Info`, `CSIStorageByClass`). IMHO this is better in the following version which adds `CSIDriver.Status`:

```
type CSIDriver struct {
	metav1.TypeMeta
	metav1.ObjectMeta

	// Specification of the CSI Driver.
	Spec CSIDriverSpec

	// Status of the CSI Driver.
	Status CSIDriverStatus
}

// CSIDriverStatus represents dynamic information about the driver and
// the storage provided by it, like for example current capacity.
type CSIDriverStatus struct {
	// Each driver can provide different kinds of storage.
	Storage []CSIStorage
}

// CSIStorage contains information for one particular kind of storage
// provided by a CSI driver.
type CSIStorage struct {
	// The name of an object is the same as a
	// StorageClass for storage accessed via that storage class,
	// <ephemeral> for storage used by ephemeral inline volumes (which
	// don't use a storage class) or <all> for storage that is the same
	// regardless of the storage class.
	StorageClassName string

	// A CSI driver may allocate storage from one or more pools
	// with different attributes.
	Pools []CSIStoragePool
}

// CSIStoragePoolInfo identifies one particular storage pool and
// stores the corresponding attributes.
type CSIStoragePool struct {
	// The name is some user-friendly
	// identifier for this entry.  It has to be unique inside the
	// CSIStorage.Pools list.
	Name string

	// NodeTopology can be used to describe a storage pool that is available
	// only for certain nodes in the cluster. If not set, the pool is consider
	// to be available from all nodes.
	NodeTopology *api.NodeSelector

	// Capacity is the size of the largest volume that currently can
	// be created. This is a best-effort guess and even volumes
	// of that size might not get created successfully.
	Capacity *resource.Quantity

	// ExpiryTime is the absolute time at which this entry becomes obsolete.
	// When not set, the entry is valid forever.
	ExpiryTime *metav1.Time
}
```